WIP: feat/request signer

vasarhelyi opened the pull request:
This PR supports communication with Skybrush Studio Gateway for request signing of pro users and progress display in gateway GUI. Main issues already resolved are summarized below:

- [x] progress reporting and in general any POST request issuing is very very slow currently. Did not debug yet why
- [x] might need to rename things yet (e.g. signer vs gateway vs license manager etc.)
- [x] error handling of progress is not really present yet
- [x] error handling of request signing could be enhanced, needs a lot of testing
- [x] documentation update
- [x] install instruction update
- [x] thorough testing with new backend and request signer

It is probably merge-ready in terms of functionality and testing, but code review is good to have. New release of Studio Blender is needed ONLY when the new account + shop site is up and running.

vasarhelyi committed:
feat: add global settings to set signer url and register HW ID at SB account

WIP: no actual signing is implemented yet

vasarhelyi committed:
feat: add request signing

needed some refactor

vasarhelyi committed:
fix: make skybrush account link URI-safe

vasarhelyi committed:
docs: updated changelog with request signing

vasarhelyi committed:
fix: fix sb account url template

vasarhelyi committed:
fix: make gzip compression in requests deterministic

vasarhelyi committed:
fix: remove fallback api key, use None

vasarhelyi committed:
fixup! fix: remove fallback api key, use None

vasarhelyi committed:
fix: small fixes to make signing work

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
fix: rename community server to studio server that is more descriptive of the future

vasarhelyi committed:
fix: make sure we sign and request the same compressed data

vasarhelyi committed:
chore: rename signer to gateway

vasarhelyi committed:
fix: enforcing post over get became a keyword param of send_request, it is only needed there

vasarhelyi committed:
feat: show progess information through Studio Gateway

vasarhelyi committed:
fix: fix gateway progress reporting

WIP: very slow yet...

ntamas commented on src/modules/sbstudio/api/gateway.py:
```diff
@@ -0,0 +1,105 @@
+    Skybrush Gateway for request signing and progress display.
+    """
+
+    _progress_task_url: str | None = None
```
This does not belong here as (in theory) you could be performing multiple progress-related operations in parallel and then you would need to maintain two such URLs.

A better pattern would be to use a context manager. I will elaborate on this idea in this PR later.

ntamas commented on src/modules/sbstudio/plugin/model/global_settings.py:
```diff
@@ -1,12 +1,43 @@
+            log.warning(
+                f"Studio gateway could not be reached at {self.gateway_url}: {ex}"
+            )
+            pass
```
No need to `pass` here

ntamas commented on src/modules/sbstudio/plugin/model/global_settings.py:
```diff
@@ -35,9 +71,19 @@ class DroneShowAddonGlobalSettings(AddonPreferences):
+        ),
+    )
+
+    gateway_url = StringProperty(
```
As we discussed earlier, the user should not be able to override the gateway URL, otherwise it is much easier to stand in as a middleman batween the gateway and the Blender plugin.

However, we need to take care of the situation when the port we have chosen for the gateway is already occupied by some other service on the system. We will probably need to implement SSDP service discovery later to allow the port of the gateway to be dynamic.

ntamas commented on src/modules/sbstudio/plugin/model/global_settings.py:
```diff
@@ -61,7 +126,7 @@ def draw(self, context):
         op.url = "http://localhost:8000"
 
-        op = row.operator(SetServerURLOperator.bl_idname, text="Use community server")
+        op = row.operator(SetServerURLOperator.bl_idname, text="Use cloud server")
```
I would keep the "community" server name. "Cloud" could mean any other machine, not only ours. "Community" makes it clear that this is the one that we maintain ourselves for the community.

vasarhelyi replied:
what about "online"? It is a nice antagonist to "local" and more general than "community"

vasarhelyi replied:
I reverted to community. Closing this issue.

ntamas commented on src/modules/sbstudio/plugin/operators/register_hardware_id.py:
```diff
@@ -0,0 +1,35 @@
+        hardware_id = prefs.hardware_id
+
+        if hardware_id:
+            open(SKYBRUSH_ACCOUNT_URL_TEMPLATE.format(hardware_id=hardware_id))
```
I think we should import the entire `webbrowser` module and then call this as `webbrowser.open()`. Otherwise it is too easy to confuse this `open()` call with the `open()` built-in function.

ntamas commented on src/modules/sbstudio/plugin/gateway.py:
```diff
@@ -0,0 +1,72 @@
+    the user does not change the add-on settings.
+
+    Raises:
+        ValueError on initialization error
```
```suggestion
        ValueError: on initialization error
```

(this is the standard syntax that can be parsed)

ntamas committed:
refactor: remove unneeded extra variables

ntamas commented on src/modules/sbstudio/api/gateway.py:
```diff
@@ -0,0 +1,105 @@
+    def get_hardware_id(self) -> str:
+        """Gets the hardware ID of the current machine from the Studio Gateway."""
+        with self._send_request("hwid") as response:
+            result = response.as_str()
```
Fixed.

ntamas commented on src/modules/sbstudio/api/gateway.py:
```diff
@@ -0,0 +1,105 @@
+        with self._send_request(
+            "sign", data=data, compressed=compressed, method="POST"
+        ) as response:
+            result = response.as_str()
```
Fixed.

ntamas committed:
fix: fixes for minor issues in latest review round

vasarhelyi commented on src/modules/sbstudio/plugin/operators/register_hardware_id.py:
```diff
@@ -0,0 +1,35 @@
+log = logging.getLogger(__name__)
+
+
+SKYBRUSH_ACCOUNT_URL_TEMPLATE = """https://account.skybrush.io/login?redirect=%2Flicenses%3FpendingHardwareId%3D{hardware_id}%26pendingProduct%3Dio.skybrush.studio.server"""
```
@isti115 Can you comment on this? I followed your instructions on creating this URL pattern.

vasarhelyi replied:
`docs\studio-licensing.md` got updated, I have rewritten blender plugin code accordingly, waiting for @isti115 to give feedback when the account site will be ready to accept the new url format.

ntamas committed:
refactor: make the internals of ProgressReport private

The caller should not be concerned with managing the timestamps in the progress report instance

ntamas committed:
fix: remove unneeded type annotation

ntamas committed:
refactor: FrameProgressReport does not need to be public, progress handlers should accept any generic ProgressReport

ntamas committed:
refactor: rename local variable "range" to "frames" as "range" is a Python built-in

ntamas committed:
refactor: remove by_name=... arguments as they complicate typing and we always used by_name=True

ntamas committed:
refactor: progress handler refactoring, WIP

ntamas committed:
feat: progress reporting through gateway now seems to work nicely

vasarhelyi committed:
fix: remove debug raise

vasarhelyi committed:
chore: get rid of Optional[]

vasarhelyi committed:
fix: make sure localhost is resolved to IPv4 on Windows, speeding up POST requests of progress and server

vasarhelyi committed:
fix: revert prev commit (cloud->community)

vasarhelyi committed:
fix: removed unused variable

ntamas committed:
fix: try to handle KeyboardInterrupt gracefully in progress display

vasarhelyi committed:
fix: update minimum required backend version

vasarhelyi committed:
fix: fix SB account's hardwareID registration link

vasarhelyi committed:
feat: handle remote task cancelling during gateway progress reporting

ntamas committed:
refactor: rename SkybrushStudioTaskCancelledError to TaskCancelled

ntamas committed:
refactor: merge TaskCancelled and KeyboardInterrupt blocks

vasarhelyi committed:
fix: fix sb account url, now works finally

vasarhelyi committed:
fix: harmonize task cancellation

vasarhelyi committed:
feat: report user cancel back to gateway's progress reporter

vasarhelyi committed:
fix: slightly better console error handler (does not make a huge diff)

vasarhelyi committed:
fix: check gateway explicitly before using it as a progress reporter

vasarhelyi committed:
docs: update install docs with gateway usage

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
docs: added screenshots from studio server and gateway

vasarhelyi committed:
docs: typo fix

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

ntamas committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
Merge tag '4.1.0' into feat/request-signer

4.1.0

vasarhelyi committed:
fixup! Merge tag '4.1.0' into feat/request-signer

vasarhelyi committed:
fixup! fixup! Merge tag '4.1.0' into feat/request-signer

vasarhelyi committed:
fixup! fixup! fixup! Merge tag '4.1.0' into feat/request-signer

vasarhelyi committed:
Merge tag '4.1.1' into feat/request-signer

4.1.1

vasarhelyi committed:
Merge tag '4.2.0' into feat/request-signer

4.2.0

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
fix: add missing import

vasarhelyi committed:
fix: slightly better error handling on unexpected errors on API calls

vasarhelyi committed:
chore: resolve some minor ty checks

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
fixup! Merge branch 'main' into feat/request-signer

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
fix: remove remainings of by_name param of sample_positions_and_colors_of_objects()

vasarhelyi committed:
Merge branch 'main' into feat/request-signer

ntamas committed:
Merge branch 'main' into feat/request-signer

vasarhelyi committed:
fix: enhance progress reporting when traj and light FPS differ

vasarhelyi committed:
fix: add FPS to progress titles to be more informative

vasarhelyi committed:
feat: implement and use suspended_color_update_callbacks() to speed up export processes regardless of color visualization types

ntamas committed:
Merge branch 'main' into feat/request-signer

ntamas committed:
fix: fix linter warnings

ntamas committed:
style: slightly tweaked the layout of the add-on settings

ntamas committed:
doc: remove trivial exception declaration from docstring

ntamas committed:
refactor: rename SkybrushStudioResponse back to Response

ntamas committed:
refactor: get rid of deprecated HTTP method calls

ntamas committed:
refactor: move the logic required to check for online access to a single
decorator

ntamas committed:
chore: typing improvements

ntamas committed:
refactor: keep ProgressReport protocol truly minimal (as promised in the
docstring)

Removed all methods from ProgressReport that are not used from the
outside. Also converted remaining_time to a method.

ntamas committed:
doc: document gateway_url_updated callback

ntamas committed:
doc: nitpickng, assure -> ensure

ntamas commented on doc/modules/ROOT/pages/install.adoc:
```diff
@@ -8,17 +8,75 @@ Please follow the steps below.
-=== 3. Install Skybrush Studio for Blender
+=== Install Skybrush Gateway
+
+*Skybrush Gateway* is a local component running on your machine, responsible for assuring trusted communication with our online server. It also shows you real-time progress information for your long lasting Blender tasks.
```
Technically, this is not true. The gateway has nothing to do with ensuring trusted communication -- that's the job of HTTPS. (After all, if the gateway would be responsible for this, then any requests made to the community server without passing through the gateway would _not_ be communicating with our servers in a trusted manner, would it?).

The gateway instead provides a cryptographic proof about the identity of the user submitting the request. We need to write this in the documentation in a non-technical manner.

ntamas commented on doc/modules/ROOT/pages/install.adoc:
```diff
@@ -8,17 +8,75 @@ Please follow the steps below.
+
+1. Download a precompiled copy of *Skybrush Gateway* from your https://account.skybrush.io/apps[Skybrush Account].
+
+2. Start your gateway instance along with starting Blender. *Skybrush Gateway* will run in the background by default. If you wish to see task progress or other useful info hosted by *Skybrush Gateway*, open it up from its tray icon.
```
> will run in the background by default

Not true; on Windows it runs in the foreground by default, right?

ntamas commented on doc/modules/ROOT/pages/install.adoc:
```diff
@@ -8,17 +8,75 @@ Please follow the steps below.
+
+2. Start your gateway instance along with starting Blender. *Skybrush Gateway* will run in the background by default. If you wish to see task progress or other useful info hosted by *Skybrush Gateway*, open it up from its tray icon.
++
+WARNING: Keep *Skybrush Gateway* running throughout your entire design process to facilitate trusted communication with our online server.
```
Same comment about "trusted communication" - it's not about trust, it's about providing information about the user's identity to the cloud server.

ntamas commented on doc/modules/ROOT/pages/install.adoc:
```diff
@@ -8,17 +8,75 @@ Please follow the steps below.
+
+3. Press the btn:[Use local gateway] and btn:[Register HWID] buttons in the "Skybrush Studio" add-on preferences in Blender to register your machine to be used with your license automatically.
+
+4. After registering your hardware ID in your *Skybrush Account*, also make sure it is assigned to the proper *Skybrush Studio Server* component you obtained a license for (online or local, depending on which one you will use).
```
> also make sure it is assigned to the proper *Skybrush Studio Server* component you obtained a license for

We should tell the user _how_ to ensure this, i.e. where, on which webpage.

ntamas commented on doc/modules/ROOT/pages/install.adoc:
```diff
@@ -8,17 +8,75 @@ Please follow the steps below.
+
+*Skybrush Studio Server* is the required backend of *Skybrush Studio for Blender*. It is responsible for running all the complicated maths behind the scenes and thus lets you use all the automated smart design tools available in the Blender plugin.
+
+*Skybrush Studio Server* is available online for all, this is called our community server. It can also be purchased to run on your machine locally.
```
Make it extremely clear (even in the title of the chapter maybe) that this is _optional_. Right now a careless user reading only the headings in the documentation from top to bottom would say "aha, okay, I need to install Blender, then install the gateway, then install the server --- wtf???"

ntamas replied:
I  think that maybe we should rewrite this entire section of the documentation such that it focuses on the three new methods of using the add-on:

* using it with the community server
* using it with Skybrush Studio Cloud (i.e. online server, but with pro features)
* using it locally

And then there's no need to branch in each section based on whether you are a community user or a pro user.

ntamas committed:
feat: add 'active' property to Suspension to avoid having to reach into
a private property

ntamas committed:
doc: explain why we need two separate suspension methods for light
effects

ntamas committed:
chore: typing clarification

ntamas committed:
refactor: decrease nesting level of context managers during export

ntamas committed:
refactor: move request signing to a separate function, alleviating the
need for a separate _sign_and_send_request() function

ntamas committed:
refactor: clearer distinction between JSON and binary payloads in
_send_request()
